modules/navier_stokes/src/fvkernels/WCNSFV2PInterfaceAreaSourceSink.C

```diff
@@ -0,0 +1,159 @@
+    _Kc(0.129),        // Model parameter - should not be tunable
+    _gamma_b(0.264),   // Model parameter - should not be tunable
+    _Kb(1.37),         // Model parameter - should not be tunable
+    _shape_factor(6.0) // Model parameter - should not be tunable
+{
+  if (_dim >= 2 && !_v_var)
+    paramError("v", "In two or more dimensions, the v velocity must be supplied!");
+
+  if (_dim >= 3 && !_w_var)
+    paramError("w", "In three or more dimensions, the w velocity must be supplied!");
+}
+
+ADReal
+WCNSFV2PInterfaceAreaSourceSink::computeQpResidual()
+{
+
+  // Useful Arguments
+  const auto state = determineState();
+  const auto old_state = Moose::StateArg(1, Moose::SolutionIterationType::Nonlinear);
+  const auto elem_arg = makeElemArg(_current_elem);
+  const bool is_transient = _subproblem.isTransient();
+
+  // Current Variables
+  const auto u = _u_var(elem_arg, state);
+  const auto rho_d = _rho_d(elem_arg, state);
+  const auto rho_d_grad = _rho_d.gradient(elem_arg, state);
+  const auto xi = _var(elem_arg, state);
+  const auto rho_m = _rho_mixture(elem_arg, state);
+  const auto f_d = _f_d(elem_arg, state);
+  const auto sigma = _sigma(elem_arg, state);
+  const auto rho_l = (rho_m - f_d * rho_d) / (1.0 - f_d + libMesh::TOLERANCE);
+  const auto complement_fd = std::max(_f_d_max - f_d, libMesh::TOLERANCE);
+  const auto f_d_o_xi = f_d / (_var(elem_arg, state) + libMesh::TOLERANCE) + libMesh::TOLERANCE;
+  const auto f_d_o_xi_old =
+      f_d / (_var(elem_arg, old_state) + libMesh::TOLERANCE) + libMesh::TOLERANCE;
+
+  // Adding bubble compressibility term
+  ADReal material_time_derivative_rho_d = u * rho_d_grad(0);
+  if (_dim > 1)
+  {
+    material_time_derivative_rho_d += (*_v_var)(elem_arg, state) * rho_d_grad(1);
+    if (_dim > 2)
+    {
+      material_time_derivative_rho_d += (*_w_var)(elem_arg, state) * rho_d_grad(2);
+    }
+  }
+  if (is_transient)
+    material_time_derivative_rho_d +=
+        (_rho_d(elem_arg, state) - _rho_d(elem_arg, Moose::oldState())) / _dt;
+  const auto bubble_compressibility = material_time_derivative_rho_d * xi / 3.0;
+
+  // Adding area growth due to added mass
+  ADReal bubble_added_mass;
+  if (f_d < _cutoff_fraction)
+    bubble_added_mass = f_d * 6.0 / _particle_diameter(elem_arg, state) - _var(elem_arg, old_state);
+  else
+    bubble_added_mass = 2. / 3. * _mass_exchange_coefficient(elem_arg, state) *
+                        (1.0 / (f_d + libMesh::TOLERANCE) - 2.0);
+
+  // // Model parameters
+  const auto db = _shape_factor * f_d_o_xi_old + libMesh::TOLERANCE;
+
+  ADRealVectorValue velocity(u);
+  if (_v_var)
+    velocity(1) = (*_v_var)(elem_arg, state);
+  if (_w_var)
+    velocity(2) = (*_w_var)(elem_arg, state);
+
+  const ADReal velocity_norm = NS::computeSpeed(velocity);
+  const auto pressure_gradient = raw_value(_pressure.gradient(elem_arg, state));
+  const ADReal pressure_grad_norm =
+      MooseUtils::isZero(pressure_gradient) ? 1e-42 : pressure_gradient.norm();
```

Comment:
If this discards derivatives, why not just real for the grad_norm? Also this might mess up the Jacobian if you are solving with the monolithic solver. Can you just lag the pressure gradient (so evaluate with `old_state`).

Reply:
Chaged according to oral discussion.